benchmarks/distributed/tensor/tp_benchmark.py

```diff
@@ -0,0 +1,247 @@
+    SequenceParallel,
+    parallelize_module,
+)
+from torch.distributed.nn.functional import (
+    _reduce_scatter_base,
+    _all_gather_base,
+    all_reduce,
+)
+from torch.utils._pytree import tree_flatten, tree_map
+
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+
+    # initialize the process group
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+class DummyDTensor(torch.Tensor):
+    _local_tensor: torch.Tensor
+
+    @staticmethod
+    def __new__(
+        cls,
+        local_tensor: torch.Tensor,
+    ) -> "DummyDTensor":
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            local_tensor.size(),
+            strides=local_tensor.stride(),
+            dtype=local_tensor.dtype,
+            device=local_tensor.device,
+            layout=local_tensor.layout,
+            requires_grad=local_tensor.requires_grad,
+        )
+        r._local_tensor = local_tensor
+        return r
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap_schema(e: object) -> object:
+            return e._local_tensor if isinstance(e, DummyDTensor) else e
+
+        def wrap_schema(e: object) -> object:
+            return DummyDTensor(e) if isinstance(e, torch.Tensor) else e
+        args_unwrap = tree_map(unwrap_schema, args)
+        kwargs_unwrap = tree_map(unwrap_schema, kwargs)
+        return tree_map(wrap_schema, func(*args_unwrap, **kwargs_unwrap))
+
+
+
+class MLPModel(nn.Module):
+    def __init__(self, dim_size):
+        super(MLPModel, self).__init__()
+        self.net1 = nn.Linear(128, dim_size)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(dim_size, 128)
+
+    def forward(self, x):
+        return self.net2(self.relu(self.net1(x)))
+
+
+
+
+@contextlib.contextmanager
+def maybe_run_profiler(args, path):
+    if args.dump_profiler:
+        with torch.profiler.profile(
+            activities=[
+                torch.profiler.ProfilerActivity.CPU,
+                torch.profiler.ProfilerActivity.CUDA,
+            ],
+            schedule=torch.profiler.schedule(wait=1, warmup=2, active=2, repeat=1),
+            on_trace_ready=torch.profiler.tensorboard_trace_handler(path),
+            profile_memory=True,
+            with_stack=True,
+            record_shapes=True,
+        ) as torch_profiler:
+            yield torch_profiler
+    else:
+        torch_profiler = contextlib.nullcontext()
+        yield torch_profiler
+
+
+def run_tp(rank, args):
+    """
+    Main body of the demo of a basic version of tensor parallel by using
+    PyTorch native APIs.
+    """
+    print(f"Running basic Megatron style TP example on rank {rank}.")
+    setup(rank, args.world_size)
+    tp_degree = args.world_size
+    dim_size = 64
```

Comment:
I am wondering if we should mimic the real world run on dim size to conduct this experiment? i.e. we should figure out the 65B model MLP matmul shape and use that to measure, because when going large the compute time will also become larger, which could possibly affect the benchmarks.

Reply:
yes, I can change to the size of MLP within 65B model.